### DISCUSSION TITLE
Support for clipboard?

### DISCUSSION THREAD

**Discussion by csterritt** [wailsapp/wails#1659] [Q&A]
Hi -- I have the following code in my JS, which works fine for copying text to the clipboard when I'm running in development, but throws an error when I try it in the production app.

~~~
    copyUrlToClipboard(url) {
      DebugOutput(`About to write to the clipboard: navigator is: '${navigator}'`)
      try {
        navigator.clipboard.writeText(url).catch((err) => {
          const msg = `Unable to copy to the clipboard: ${err}`
          DebugOutput(`Error: ${msg}`)
          LogPrint(msg)
          console.log(msg)
        })
      } catch (err) {
        const msg = `Unable to copy to the clipboard (in try/catch): ${err}`
        DebugOutput(`Error: ${msg}`)
        LogPrint(msg)
        console.log(msg)
      }
      DebugOutput(`Finished write to the clipboard`)
    },
~~~

I get a complaint that `navigator.clipboard is undefined` at runtime, but again, only in the production app.

Any clues? Thanks!

**Comment by leaanthony** [comment]
When you say in Dev mode, do you mean the desktop app or browser?

**Reply by csterritt** [reply]
Yes :-)

When I run `wails dev`, either the desktop app or browser works to do the copy.

When I run `wails build`, the application built works fine, but I get the above error.

**Reply by leaanthony** [reply]
Oh... Well that is weird as there's little to no difference between the app in Dev mode and the app in prod mode...

**Reply by leaanthony** [reply]
When I get home I'll convert this to an issue.

**Reply by csterritt** [reply]
I wonder if there's a way to have the Go side make a system call to copy the object/text?

I say this because the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard) says, "Secure context: This feature is available only in secure contexts (HTTPS)". Which just seems to add to the mystery -- does WebView somehow set up a certificate on the server side to make its browser think it's secure? Then why doesn't it do it in production?

**Reply by leaanthony** [reply]
What's more, webview runs in exactly the same way in prod builds as it does in Dev....

**Reply by ianmjones** [reply]
I do clipboard read/write from Go rather than JS side of things using the following package:

https://pkg.go.dev/golang.design/x/clipboard

**Reply by csterritt** [reply]
@ianmjones THANK YOU! That works perfectly, in both dev and prod.

Maybe this should part of the Wails Go API?

**Reply by leaanthony** [reply]
Maybe.... At the very least it would be good to add a guide in the docs website for this. It would be an amazing way to contribute to the project @csterritt 😀
